### PR TITLE
fix(mute alerts): Mute digests

### PR DESCRIPTION
**Pull request by ceorourke**
Make sure that we don't send digest notifications for rules that have been muted.

### CODE REVIEW

File: src/sentry/digests/utils.py [hunk 1]
```diff
@@ -132,6 +133,7 @@ def get_participants_by_event(
             target_identifier=target_identifier,
             event=event,
             fallthrough_choice=fallthrough_choice,
+            rules=rules,
```
Comment (ceorourke) [inline src/sentry/digests/utils.py:136]: I think the only place that we weren't sending `rules` is for digests so I could rewrite [get_send_to](https://github.com/getsentry/sentry/blob/master/src/sentry/notifications/utils/participants.py#L391) a bit to expect it, though I have to update a bunch of tests as well. Might be better suited for a separate PR.

Comment (roggenkemper) [reply src/sentry/digests/utils.py:136]: one thing we should probably do when we re-do get_send_to is that right now we check if any of the rules have a snooze for all , then no notification gets sent which I don't think works for digests. It might be trickier because for digests I think we will have to change what gets sent in the notification, not just if it gets sent or not

Comment (ceorourke) [reply src/sentry/digests/utils.py:136]: Ohhh yeah that's a good point ... that should probably be addressed now.

File: src/sentry/digests/utils.py [hunk 2]
```diff
@@ -109,8 +122,10 @@ def build_custom_digest(original_digest: Digest, events: Iterable[Event]) -> Dig
             if user_group_records:
                 user_rule_groups[group] = user_group_records
         if user_rule_groups:
-            user_digest[rule] = user_rule_groups
+            if not skip:
```
Comment (roggenkemper) [inline src/sentry/digests/utils.py:125]: will skip ever be `True` here? if it is set to True in the inner for loop, won't it continue to the next iteration in the outer loop and reset skip to False?

Comment (ceorourke) [reply src/sentry/digests/utils.py:125]: oh yeah woops reset it at the wrong spot

File: src/sentry/digests/utils.py [hunk 3]
```diff
@@ -97,10 +99,20 @@ def get_event_from_groups_in_digest(digest: Digest) -> Iterable[Event]:
...
 
-def build_custom_digest(original_digest: Digest, events: Iterable[Event]) -> Digest:
+def build_custom_digest(
+    original_digest: Digest, events: Iterable[Event], participant: RpcActor
+) -> Digest:
     """Given a digest and a set of events, filter the digest to only records that include the events."""
     user_digest: Digest = {}
     for rule, rule_groups in original_digest.items():
+        skip = False
+        rule_snoozes = RuleSnooze.objects.filter(Q(rule=rule))
```
Comment (roggenkemper) [inline src/sentry/digests/utils.py:109]: do we need the `Q` here?

File: src/sentry/digests/utils.py [hunk 4]
```diff
@@ -97,10 +99,20 @@ def get_event_from_groups_in_digest(digest: Digest) -> Iterable[Event]:
...
+        for snooze in rule_snoozes:
+            if snooze.user_id is None or snooze.user_id == participant.id:
+                skip = True
+                break
+        if skip:
+            continue
```
Comment (wedamija) [inline src/sentry/digests/utils.py:115]: Nit: This is a case for a `for ... else`

``` 
for snooze in rule_snoozes:
    if snooze.user_id is None or snooze.user_id == participant.id:
        break
else:
    continue
```

Doesn't really matter it's just rare that to actually have a use for it. The else runs if we never break.

File: src/sentry/digests/utils.py (hunk 3 above)
Comment (wedamija) [inline src/sentry/digests/utils.py:109]: Could be worth batch fetching these before the for loop to avoid n+1 queries

File: src/sentry/digests/utils.py [hunk 5]
```diff
@@ -97,10 +99,20 @@ def get_event_from_groups_in_digest(digest: Digest) -> Iterable[Event]:
...
+def build_custom_digest(
+    original_digest: Digest, events: Iterable[Event], participant: RpcActor
+) -> Digest:
     """Given a digest and a set of events, filter the digest to only records that include the events."""
     user_digest: Digest = {}
     for rule, rule_groups in original_digest.items():
+        skip = False
+        rule_snoozes = RuleSnooze.objects.filter(Q(rule=rule))
+        for snooze in rule_snoozes:
+            if snooze.user_id is None or snooze.user_id == participant.id:
```
Comment (wedamija) [inline src/sentry/digests/utils.py:111]: Might actually make sense to include this clause in your query criteria. Then if there are any `rule_snooze` rows you can just continue.

File: src/sentry/digests/utils.py (hunk 4 above)
Comment (ceorourke) [reply src/sentry/digests/utils.py:118]: 🤔 I tried this out and while it does break the inner loop of the snooze, it just continues to add stuff to the digest that I don't want

Comment (wedamija) [reply src/sentry/digests/utils.py:118]: Oh I think I got it backwards... Yeah, I don't think it'll work. Just got overexcited to use a for/else

Comment (ceorourke) [reply src/sentry/digests/utils.py:118]: Lol I'll try to sneak it in somewhere, someday, and when I do, I'll tag you.

**Review by roggenkemper — APPROVED** [review]

File: src/sentry/digests/utils.py [hunk 6]
```diff
@@ -97,10 +104,23 @@ def get_event_from_groups_in_digest(digest: Digest) -> Iterable[Event]:
...
 
 
-def build_custom_digest(original_digest: Digest, events: Iterable[Event]) -> Digest:
+def build_custom_digest(
+    original_digest: Digest, events: Iterable[Event], participant: RpcActor
+) -> Digest:
     """Given a digest and a set of events, filter the digest to only records that include the events."""
     user_digest: Digest = {}
+    rule_snoozes = RuleSnooze.objects.filter(
+        Q(Q(user_id=participant.id) | Q(user_id__isnull=True)), rule__in=original_digest.keys()
```
Comment (wedamija) [inline src/sentry/digests/utils.py:113]: Nit: Btw I realise I lied, you don't need the outer `Q` here, since it's a separate argument to filter anyway

File: src/sentry/digests/utils.py [hunk 7]
```diff
@@ -97,10 +104,23 @@ def get_event_from_groups_in_digest(digest: Digest) -> Iterable[Event]:
...
+    )
+
     for rule, rule_groups in original_digest.items():
+        skip = False
+        for snooze in rule_snoozes:
+            if snooze.rule == rule:
+                skip = True
+                break
```
Comment (wedamija) [inline src/sentry/digests/utils.py:121]: Nit: Now that you have all the rule snoozes in one place, you could just store them as a set of `snoozed_rule_ids`, then say `if rule.id in snoozed_rule_ids: continue`

File: src/sentry/notifications/notifications/digest.py [hunk 8]
```diff
@@ -193,8 +193,17 @@ def send(self) -> None:
...
             extra_context = self.get_extra_context(participants_by_provider_by_event)
 
         for provider, participants in combined_participants_by_provider.items():
+            if personalized_digests:
+                # remove participants if the digest is empty
+                participants_to_remove = set()
+                for participant in participants:
+                    if participant.actor_id not in set(extra_context.keys()):
+                        participants_to_remove.add(participant)
+                participants -= participants_to_remove
```
Comment (wedamija) [inline src/sentry/notifications/notifications/digest.py:208]: Should we just remove from participants directly, rather than adding to the other set first then subtracting?

File: src/sentry/notifications/notifications/digest.py [hunk 9]
```diff
@@ -193,8 +193,17 @@ def send(self) -> None:
...
+
+        if personalized_digests:
             extra_context = self.get_extra_context(participants_by_provider_by_event)
 
         for provider, participants in combined_participants_by_provider.items():
+            if personalized_digests:
+                # remove participants if the digest is empty
+                participants_to_remove = set()
+                for participant in participants:
+                    if participant.actor_id not in set(extra_context.keys()):
```
Comment (wedamija) [inline src/sentry/notifications/notifications/digest.py:206]: You can use `participant.actor_id not in extra_context`. That'll look up by the keys anyway and will be more efficient, since the current way requires us to enumerate all the keys and convert to a set first.

**Review by wedamija — APPROVED** [review]
lgtm, rest of the comments are just nits

File: src/sentry/notifications/notifications/digest.py (hunk 8 above)
Comment (ceorourke) [reply src/sentry/notifications/notifications/digest.py:208]: I tried that and got an error about the set changing size during iteration

Comment (wedamija) [reply src/sentry/notifications/notifications/digest.py:208]: Ahh fair enough, yeah leave it as is then.
